lib/livebook_web/plugs/auth_plug.ex

```diff
@@ -64,10 +73,18 @@ defmodule LivebookWeb.AuthPlug do
       |> redirect(to: path_with_query(conn.request_path, query_params))
       |> halt()
     else
-      redirect_to_authenticate(conn)
+      conn
+      |> put_redirect_path_in_session()
+      |> redirect_to_authenticate()
     end
   end
 
+  defp put_redirect_path_in_session(
+         %{request_path: request_path, query_params: query_params} = conn
+       ) do
+    put_session(conn, :redirect_to, path_with_query(request_path, query_params))
+  end
+
   defp redirect_to_authenticate(conn) do
     conn
     |> redirect(to: "/authenticate")
```

Comment:
The logic is the same for both password and token so we can move it here :)

```suggestion
  defp redirect_to_authenticate(conn) do
    conn
    |> put_session(:redirect_to, path_with_query(conn.request_path, conn.query_params))
    |> redirect(to: "/authenticate")
```